### PR TITLE
[SPARK-32920][FOLLOW-UP] Fix shuffleMergeFinalized directly calling rdd.getNumPartitions as RDD is not serialized to executor

### PR DESCRIPTION
**Pull request by venkata91**
### What changes were proposed in this pull request?

`ShuffleMapTask` should not push blocks if a shuffle is already merge finalized. Currently block push is disabled for retry cases. Also fix `shuffleMergeFinalized` calling `rdd.getNumPartitions` as RDD is not serialized causing issues.

### Why are the changes needed?

No

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing tests

### CODE REVIEW

**Comment by venkata91** [comment]
cc @otterc @mridulm Please take a look.

**Comment by otterc** [comment]
cc @Ngone51

File: core/src/main/scala/org/apache/spark/shuffle/ShuffleWriteProcessor.scala [hunk 1]
```diff
@@ -64,7 +64,8 @@ private[spark] class ShuffleWriteProcessor extends Serializable with Logging {
         // The map task only takes care of converting the shuffle data file into multiple
         // block push requests. It delegates pushing the blocks to a different thread-pool -
         // ShuffleBlockPusher.BLOCK_PUSHER_POOL.
-        if (Utils.isPushBasedShuffleEnabled(SparkEnv.get.conf) && dep.getMergerLocs.nonEmpty) {
+        if (Utils.isPushBasedShuffleEnabled(SparkEnv.get.conf) && dep.getMergerLocs.nonEmpty &&
+            !dep.shuffleMergeFinalized) {
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/shuffle/ShuffleWriteProcessor.scala:68]: What happens if the push isn't finalized in the last stage attempt but the task itself is from a new stage attempt?

Comment (venkata91) [reply core/src/main/scala/org/apache/spark/shuffle/ShuffleWriteProcessor.scala:68]: Then the shuffle dependency won't be `finalized` therefore the new task from a new stage attempt would be able to push.

Comment (mridulm) [reply core/src/main/scala/org/apache/spark/shuffle/ShuffleWriteProcessor.scala:68]: Remove `Utils.isPushBasedShuffleEnabled(SparkEnv.get.conf)` ?

File: core/src/main/scala/org/apache/spark/Dependency.scala [hunk 2]
```diff
@@ -97,6 +97,8 @@ class ShuffleDependency[K: ClassTag, V: ClassTag, C: ClassTag](
   val shuffleHandle: ShuffleHandle = _rdd.context.env.shuffleManager.registerShuffle(
     shuffleId, this)
 
+  val numPartitions = rdd.partitions.length
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/Dependency.scala:100]: `private[this]` ?

File: core/src/main/scala/org/apache/spark/shuffle/ShuffleWriteProcessor.scala (hunk 1 above)
Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/shuffle/ShuffleWriteProcessor.scala:68]: > Then the shuffle dependency won't be finalized therefore the new task from a new stage attempt would be able to push.

@venkata91  So, would there be data correctness issue on the merger side (ESS)? Who's in charge of the data file maintains on the merger side in the case of the stage rerun? (I think this's a different issue with indeterministic stage return)

Comment (Ngone51) [reply core/src/main/scala/org/apache/spark/shuffle/ShuffleWriteProcessor.scala]: sgtm

**Comment by Ngone51** [comment]
```
[error] /home/runner/work/spark/spark/core/src/main/scala/org/apache/spark/shuffle/ShuffleWriteProcessor.scala:24:30: Unused import
[error] import org.apache.spark.util.Utils
[error]                              ^
[error] one error found
```

**Comment by Ngone51** [comment]
LGTM if tests pass.

**Comment by venkata91** [comment]
> ```
> [error] /home/runner/work/spark/spark/core/src/main/scala/org/apache/spark/shuffle/ShuffleWriteProcessor.scala:24:30: Unused import
> [error] import org.apache.spark.util.Utils
> [error]                              ^
> [error] one error found
> ```

Sorry my bad, fixed it.

**Comment by venkata91** [comment]
It seems like the build is failing with the below msg which doesn't seem to be related to this PR. cc @Ngone51 

```/home/runner/work/spark/spark/build/sbt-launch-lib.bash: line 77:  5227 Killed                  "$@"
Error:  running /home/runner/work/spark/spark/build/sbt -Phadoop-3.2 -Phive-2.3 launcher/test avro/test network-common/test sketch/test unsafe/test examples/test repl/test core/test graphx/test kvstore/test network-shuffle/test ; received return code 137
Error: Process completed with exit code 18.```

**Comment by mridulm** [comment]
Can you update to latest and rerun the tests ?

**Comment by venkata91** [comment]
> Can you update to latest and rerun the tests ?

Done

**Comment by mridulm** [comment]
Merged to master/branch-3.2 +CC @gengliangwang
Thanks for fixing this @venkata91 !
Thanks for the review @Ngone51 :-)

**Review by gengliangwang — COMMENTED** [review]
Late LGTM.
Thanks for the ping, @mridulm

**Pull request closed without merging**